RGB Animation documentation on reducing memory footprint (#5813)
qmk/qmk_firmware: docs/feature_rgblight.md

<img src="gitbook/images/color-wheel.svg" alt="HSV Color Wheel" width="250"/>

> Changing the **Hue** cycles around the circle.

Reviewer: These lines should keep the double spaces at the end, having it all on one line doesn't look as good.

Author: Sorry editor removed those as extra white space. I'll fix it

Author: I've introduced <br> as this is a better solution than space which is really relying on a quirk in markdown to introduce a line break and is invisible.

Reviewer: I thought that was a feature of markdown... https://daringfireball.net/projects/markdown/syntax#p

Author: don't disagree, but it's hard to tell that from a mistake while editing. Soooo I'd rather just use <br> and remove the ambiguity.

|`RGBLIGHT_EFFECT_BREATHE_MAX`       |`255`        |The maximum brightness for the breathing mode. Valid values are 1 to 255             |
|`RGBLIGHT_EFFECT_CHRISTMAS_INTERVAL`|`1000`       |How long to wait between light changes for the "Christmas" animation, in milliseconds|
|`RGBLIGHT_EFFECT_CHRISTMAS_STEP`    |`2`          |The number of LEDs to group the red/green colors by for the "Christmas" animation    |
|`RGBLIGHT_EFFECT_KNIGHT_LED_NUM`    |`RGBLED_NUM` |The number of LEDs to have the "Knight" animation travel                             |
|`RGBLIGHT_EFFECT_KNIGHT_LENGTH`     |`3`          |The number of LEDs to light up for the "Knight" animation                            |
|`RGBLIGHT_EFFECT_KNIGHT_OFFSET`     |`0`          |The number of LEDs to start the "Knight" animation from the start of the strip by    |
|`RGBLIGHT_RAINBOW_SWIRL_RANGE`      |`360`        |Range adjustment for the rainbow swirl effect to get different swirls                |
|`RGBLIGHT_EFFECT_SNAKE_LENGTH`      |`4`          |The number of LEDs to light up for the "Snake" animation                             |

> |`RGBLIGHT_EFFECT_STATIC_GRADIENT`   |*Not defined*|Enable static gradient mode.                                             |

Reviewer 2: Although RGBLIGHT_EFFECT_STATIC_GRADIENT is not an animation, this mode can also be suppressed, which has the effect of saving size.

### Example Usage to Reduce Memory Footprint
  1. Remove `RGBLIGHT_ANIMATIONS` from `config.h`.
  1. Selectively add the animations you want to enable. The following would enable two animations and save about 4KiB:

```diff
 #undef RGBLED_NUM
-#define RGBLIGHT_ANIMATIONS

> +#define RGBLIGHT_EFFECT_STATIC_LIGHT

Reviewer 2: RGBLIGHT_MODE_STATIC_LIGHT is always enabled and can not be disabled. Therefore, it can not be controlled by the symbol RGBLIGHT_EFFECT_STATIC_LIGHT.

Did you know this and purposely wrote this?

Author: No, it was randomly chosen. Looks like documentation is wrong. I just copied it.

Author: What do you suggest @mtei ?

Reviewer 2: I think that RGBLIGHT_MODE_STATIC_GRADIENT is good.  Because it has the most submodes. So I recommend RGBLIGHT_EFFECT_STATIC_GRADIENT.

|Define                              |Default      |Description                                                                          |
|------------------------------------|-------------|-------------------------------------------------------------------------------------|

> |`RGBLIGHT_ANIMATIONS`               |*Not defined*|Enable all additional animation modes                                   |

Reviewer: ```suggestion
|`RGBLIGHT_ANIMATIONS`               |*Not defined*|Enable all additional animation modes.                                   |
```

|`RGBLIGHT_EFFECT_KNIGHT_OFFSET`     |`0`          |The number of LEDs to start the "Knight" animation from the start of the strip by    |
|`RGBLIGHT_EFFECT_RAINBOW_MOOD`      |*Not defined*|Enable rainbow mood animation mode.                                      |
|`RGBLIGHT_EFFECT_RAINBOW_SWIRL`     |*Not defined*|Enable rainbow swirl animation mode.                                     |
|`RGBLIGHT_RAINBOW_SWIRL_RANGE`      |`360`        |Range adjustment for the rainbow swirl effect to get different swirls                |
|`RGBLIGHT_EFFECT_RGB_TEST`          |*Not defined*|Enable RGB test animation mode.                                          |
|`RGBLIGHT_EFFECT_SNAKE`             |*Not defined*|Enable snake animation mode.                                             |
|`RGBLIGHT_EFFECT_SNAKE_LENGTH`      |`4`          |The number of LEDs to light up for the "Snake" animation                             |
|`RGBLIGHT_EFFECT_STATIC_GRADIENT`   |*Not defined*|Enable static gradient mode.                                             |

### Example Usage to Reduce Memory Footprint

>   1. remove RGBLIGHT_ANIMATION from `config.h`

Reviewer: ```suggestion
  1. Remove `RGBLIGHT_ANIMATION` from `config.h`.
```

>   1. selectively add the animations you want to enable. This would enable 2 animations and save about 4K.

Reviewer: ```suggestion
  1. Selectively add the animations you want to enable. The following would enable two animations and save about 4kB.
```

Author: Maybe 4KiB to stay So units?

>   1. Selectively add the animations you want to enable. The following would enable two animations and save about 4KiB:

Reviewer: ```suggestion
  2. Selectively add the animations you want to enable. The following would enable two animations and save about 4KiB:
```

Author: No, I would let markdown do it's job. If you make that a 2 it makes reordering annoying.

Reviewer: Oh, I assumed it was the browser doing that. Carry on.

|Define                              |Default      |Description                                                                          |
|------------------------------------|-------------|-------------------------------------------------------------------------------------|
|`RGBLIGHT_ANIMATIONS`               |*Not defined*|Enable all additional animation modes.                                   |
|`RGBLIGHT_EFFECT_ALTERNATING`       |*Not defined*|Enable alternating animation mode.                                       |
|`RGBLIGHT_EFFECT_BREATHING`         |*Not defined*|Enable breathing animation mode.                                         |
|`RGBLIGHT_EFFECT_BREATHE_CENTER`    |*Not defined*|If defined, used to calculate the curve for the breathing animation. Valid values are 1.0 to 2.7 |
|`RGBLIGHT_EFFECT_BREATHE_MAX`       |`255`        |The maximum brightness for the breathing mode. Valid values are 1 to 255             |

> |`RGBLIGHT_EFFECT_CHRISTMAS`         |*Not defined*|Enable christmas animation mode.                                         |

Reviewer 3: I'm not sure about this change. I'd much rather see all the animation modes listed together.  

But maybe we should break up this table into two parts.... one for the animation modes and one for the animation settings.

Author: yeah, it's definitely a trade off. Co-locating them makes finding the info easier at the cost of obscuring the keys to enable and disable them.

Reviewer 3: Right now, it's harder to find them to turn them on, IMO.   Especially if you want just copy the list and edit that.

But honestly, I don't think there is a good solution here, that is perfect.
